clang/lib/DPCT/ExprAnalysis.cpp

```diff
@@ -584,6 +584,10 @@ void ExprAnalysis::analyzeExpr(const CXXTemporaryObjectExpr *Temp) {
        StringRef(TypeName).startswith("cooperative_groups::")) {
     analyzeType(Temp->getTypeSourceInfo()->getTypeLoc());
   }
+  if (TypeName == "__half" || TypeName == "__half2") {
```

Comment:
Adding special code in ExprAnalysis.cpp is a not good way.
In ExprAnalysis.cpp, we hope that the code can be used to solve a type of problem.
Can we analyze why it cannot be constuct/convert in the kernel parameters?

Reply:
If I delete these 2 `if` and call `analyzeType` anyway. the `ctad` case will failed because in `ctad` mode, we should not infer template argument.

Reply:
fixed